impl/src/main/java/impl/UserAnswerResourceImpl.java

```diff
@@ -43,27 +52,58 @@ public UserAnswerResourceImpl( AnswerDao answerDao) {
+                    return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
+                }
+                return answerDao.updateAnswer(auth.getUserId(), answerId, answer)
+                    .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_UPDATE_ANSWER, throwable)));
+            });
     }
 
     @Override
     public Observable<Void> deleteAnswer(Auth auth, long answerId) {
         return answerDao.getAnswerById(answerId)
             .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_GET_ANSWER_FROM_DATABASE, throwable)))
-            .switchIfEmpty(error(new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
+            .switchIfEmpty(exception(() -> new WebException(HttpResponseStatus.NOT_FOUND, ANSWER_NOT_FOUND)))
             .flatMap(storedAnswer -> {
-            if(auth.getUserId() != storedAnswer.getUserId()) {
-                return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
-            }
-            return answerDao.deleteAnswer(auth.getUserId(), answerId)
-                .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_DELETE_ANSWER, throwable)));
-        });
+                if (auth.getUserId() != storedAnswer.getUserId()) {
+                    return error(new WebException(HttpResponseStatus.FORBIDDEN, NOT_OWNER_OF_ANSWER));
+                }
+                return answerDao.deleteAnswer(auth.getUserId(), answerId)
+                    .onErrorResumeNext(throwable -> error(new WebException(INTERNAL_SERVER_ERROR, FAILED_TO_DELETE_ANSWER, throwable)));
+            });
+    }
+
+    @Override
+    public Observable<Void> upVoteAnswer(Auth auth, long answerId) {
+        return handleVote(new Vote(auth.getUserId(), answerId, 1));
+    }
+
+    @Override
+    public Observable<Void> downVoteAnswer(Auth auth, long answerId) {
+        return handleVote(new Vote(auth.getUserId(), answerId, -1));
+    }
+
+    private Observable<Void> handleVote(Vote newVote) {
+
+        return first(answerDao.getAnswerById(newVote.getAnswerId()).flatMap(answer -> {
+                if (answer.getUserId() == newVote.getUserId()) { // no voting for your own answer
+                    return error(new WebException(BAD_REQUEST, INVALID_VOTE));
+                }
+                return voteDao.findVote(newVote.getUserId(), newVote.getAnswerId());
+            })
+            .flatMap(existingVote -> {
+                int totalVote = newVote.getValue() + existingVote.getValue();
+                if (totalVote == 0) {
+                    return voteDao.deleteVote(existingVote.getUserId(), existingVote.getAnswerId())
+                        .map(i -> existingVote);
+                }
+                return error(new WebException(BAD_REQUEST, INVALID_VOTE));
+            })
+            .switchIfEmpty(voteDao.createVote(newVote).map(i -> newVote))
```

Comment:
maybe add error handling on the update operations at least?

Reply:
There is still some exception mapping in place that will result in 500 on error? I'm thinking it might be superfluous to have specific handling of unexpected errors.